CONTRIBUTING.md

```diff
@@ -40,47 +40,54 @@ standard [GitHub workflow][workflow-link]:
 1. You create a [fork][fork-link] of [google-cloud-cpp][repo-link]. Then
    [clone][about-clone] that fork into your workstation:
    ```console
-   cd $HOME
    git clone git@github.com:YOUR-USER-NAME/google-cloud-cpp.git
    ```
-2. You pick an existing [GitHub bug][mastering-issues] to work on.
-3. You start a new [branch][about-branches] for each feature (or bug fix).
+1. The cloned repo that you created in the previous step will have its `origin`
+   set to your forked repo. You should now tell git about the the main
+   `upstream` repo, which you'll use to pull commits made by others in order to
+   keep your local repo and fork up to date.
+   ```console
+   git remote add upstream git://github.com/GoogleCloudPlatform/google-cloud-cpp.git
+   git remote -v  # Should show 'origin' (your fork) and 'upstream' (main repo)
+   ```
+1. To pull new commits from `upstream` into your local repo and
+   [sync your fork][syncing-a-fork] you can do the following:
+   ```console
+   git checkout master
+   git pull upstream master
```

Comment:
Consider adding `--ff-only` to enforce that your local `master` branch is clean.

Also you can connect the two so that it doesn't `push` if `pull` fails:

```
git pull upstream master --ff-only && git push
```

Here's what I have in `~/.gitconfig` for this:

```
[alias]
  puma = !git checkout master && git pull --ff-only upstream master && git push
```

so I just write:

```
$ git puma
```

and it does the trick for me.

Reply:
Thanks for the alias suggestion.

@coryan thoughts on the `--ff-only` part? While I imagine local master should be clean, I'm not sure the pull should fail if not. But I'm happy to defer to Carlos's thoughts here.

Reply:
If you don't do this, when you create a new feature branch and push it, it'll look funny because you'll either have some extra local patches on `master` that you shouldn't, or you'll have merge commits or some other complexity, none of which are desirable.

I'm also curious to hear @coryan's thoughts on this.

Reply:
That sounds like a good practice.  It is trivial to mess up your `master`, you can always repair it, but better to prevent the problem.

Reply:
SGTM. Done.

Reply:
Right, if you checkout a feature branch when master isn't synchronized you will get an unnecessary merge commit, but I'm not sure if `--ff-only` is the fix here. Generally `git pull rebase=preserve` by default and `git checkout master; git fetch upstream; git rebase upstream/master; git checkout feature` will solve the problem.

Reply:
When starting work on a new feature, I typically run:

```
$ git puma   # see alias above
$ git checkout -b new-feature-branch
```

to make sure it's starting from a clean slate. If I've started a feature branch from an older version of `master`, or if `master` has been updated in the meantime, since I know my `master` is always a pristine clean copy of `upstream/master`, I can just run:

```
$ git puma
$ git checkout -b new-feature-branch
$ git rebase master
# fix any conflicts
$ git push -f
```

which avoids having any merge commits anywhere, and keeps history clean and easy to review.